### PR TITLE
Extend mypy checking, fix #2194

### LINKED ISSUE #2194: Extend mypy checking
A somewhat smaller amount of work than what was needed to cover the mitmproxy directory. I'd like to get us to the point where we have mypy checking for the entire codebase. 

  - [x] mitmproxy
  - [x] pathod
  - [x] examples/simple
  - [ ] examples/complex
  - [ ] examples/pathod

### PR DESCRIPTION
**Pull request by kira0204**
Fixes #2194

### CODE REVIEW

File: examples/complex/xss_scanner.py [hunk 1]
```diff
@@ -202,20 +201,21 @@ def get_SQLi_data(new_body: str, original_body: str, request_URL: str, injection
         "Sybase": (r"(?i)Warning.*sybase.*", r"Sybase message", r"Sybase.*Server message.*"),
     }
     for dbms, regexes in DBMS_ERRORS.items():
-        for regex in regexes:
+        for regex in regexes:  # type: ignore
```
Comment (mhils) [inline examples/complex/xss_scanner.py]: Why is that required?

File: examples/complex/xss_scanner.py [hunk 2]
```diff
@@ -202,20 +201,21 @@ def get_SQLi_data(new_body: str, original_body: str, request_URL: str, injection
...
                 return SQLiData(request_URL,
                                 injection_point,
                                 regex,
                                 dbms)
+    return None
 
 
 # A qc is either ' or "
-def inside_quote(qc: str, substring: bytes, text_index: int, body: bytes) -> bool:
+def inside_quote(qc: str, substring_bytes: bytes, text_index: int, body_bytes: bytes) -> bool:
```
Comment (mhils) [inline examples/complex/xss_scanner.py]: That's not really required IMO, we already have a type annotation here.

File: examples/complex/xss_scanner.py [hunk 3]
```diff
@@ -380,20 +380,21 @@ def handle_starttag(self, tag, attrs):
...
             return None
+    return None
 
 
 # response is mitmproxy's entry point
 def response(flow: http.HTTPFlow) -> None:
     cookiesDict = get_cookies(flow)
     # Example: http://xss.guru/unclaimedScriptTag.html
-    find_unclaimed_URLs(flow.response.content, flow.request.url)
+    find_unclaimed_URLs(flow.response.content.decode('utf-8'), flow.request.url)
```
Comment (mhils) [inline examples/complex/xss_scanner.py]: This will fail for any non-utf8 contents, take a look at `response.get_text` :)

File: examples/complex/xss_scanner.py (hunk 2 above)
Comment (kira0204) [reply examples/complex/xss_scanner.py]: It's not to explicitly tell about the type but earlier `substring` variable was being used to store `bytes` type and after decoding it was being used to store `str`. Thus for a better approach, I renamed this variable

File: examples/complex/xss_scanner.py (hunk 1 above)
Comment (kira0204) [reply examples/complex/xss_scanner.py]: Because mypy was throwing an error. More about this can be read in https://github.com/python/mypy/issues/2598 . I thought this would be the best solution as of now

File: examples/complex/xss_scanner.py (hunk 3 above)
Comment (kira0204) [reply examples/complex/xss_scanner.py]: Hi, we need to rewrite most of the code since there are many places where we are taking that the content will be utf8 only. `response` function has been written on this assumption so is most of the code. I thought this assumption was wrong but since this was an already existing code didn't ask.
@mhils I will go forward and check all of the code, it would be great if you could review the whole existing code once. Thank you

File: examples/complex/xss_scanner.py [hunk 4]
```diff
@@ -202,20 +201,21 @@ def get_SQLi_data(new_body: str, original_body: str, request_URL: str, injection
...
 
 # A qc is either ' or "
-def inside_quote(qc: str, substring: bytes, text_index: int, body: bytes) -> bool:
+def inside_quote(qc: str, substring_bytes: bytes, text_index: int, body_bytes: bytes) -> bool:
     """ Whether the Numberth occurence of the first string in the second
         string is inside quotes as defined by the supplied QuoteChar """
-    substring = substring.decode('utf-8')
-    body = body.decode('utf-8')
+    substring = substring_bytes.decode('utf-8')
+    body = body_bytes.decode('utf-8')
```
Comment (kira0204) [inline examples/complex/xss_scanner.py]: As you can see here, therefore used `substring_bytes`

**Comment by kira0204** [comment]
@mhils Is there anything else that needs to be done? :)

File: examples/complex/xss_scanner.py (hunk 4 above)
Comment (mhils) [reply examples/complex/xss_scanner.py]: Makes sense, thanks.

File: examples/complex/xss_scanner.py [hunk 5]
```diff
@@ -380,33 +383,38 @@ def handle_starttag(self, tag, attrs):
...
         else:
             return None
+    return None
 
 
 # response is mitmproxy's entry point
 def response(flow: http.HTTPFlow) -> None:
     cookiesDict = get_cookies(flow)
+    try:
+        resp = flow.response.content.decode('utf-8')
```
Comment (mhils) [inline examples/complex/xss_scanner.py]: Again, you assume that the response content is utf8, which is not guaranteed to be true. In fact It's latin-1 by default. Use `.get_text(strict=False)` here.

**Comment by kira0204** [comment]
Hi, @mhils Simplifying mypy invocations will need `__init__.py` , without that the build is failing because `mypy` can't find .py files

**Comment by mhils** [comment]
Thanks! 🍰

**Comment by kira0204** [comment]
Thankyou so much for guiding me :)
